components/browser/state/src/main/java/mozilla/components/browser/state/middleware/InProgressMediaMiddleware.kt

```diff
@@ -0,0 +1,44 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.state.middleware
+
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.EngineAction
+import mozilla.components.browser.state.action.MediaSessionAction
+import mozilla.components.browser.state.action.UpdateInProgressMediaTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.concept.engine.mediasession.MediaSession
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+
+/**
+ * [Middleware] that updates [BrowserState.inProgressMediaTabId] everytime the user start playing media
+ * or navigates to another page in the last tab that was playing media.
+ */
+class InProgressMediaMiddleware : Middleware<BrowserState, BrowserAction> {
+    @Suppress("ComplexCondition")
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction
+    ) {
+        next(action)
+
+        // Only interested if playing media started in a normal tab
+        if (action is MediaSessionAction.UpdateMediaPlaybackStateAction &&
+            action.playbackState == MediaSession.PlaybackState.PLAYING &&
+            context.state.tabs.firstOrNull {
+                !it.content.private && it.id == action.tabId
+            } != null) {
+
+            context.dispatch(UpdateInProgressMediaTab(action.tabId))
+        } else if (action is EngineAction.LoadUrlAction) {
+            // Forget about in progress media if the user navigates to another page (in the same session).
+            if (action.tabId == context.state.inProgressMediaTabId) {
+                context.dispatch(UpdateInProgressMediaTab(null))
+            }
```

Comment:
This feels hacky too, shouldn't this be based on playing state?

Reply:
We could check for `MediaSessionAction.DeactivatedMediaSessionAction` that changes the `MediaSessionState` to null when the user navigates to another page.

Reply:
We're gonna need to experiment with different sites and see what works best, but @pocmo pointed me to various cases (e.g. soundcloud where the same song keeps playing even when the URL changes). Let's start with a combination of `UpdateUrlAction` (not `LoadUrlAction`) AND the media being stopped/deactivated to reset the state (set the timestamp to 0).

Reply:
I think we can use just `MediaSessionAction.DeactivatedMediaSessionAction` to know when the user navigates to another page so that media stops playing. We can then also forget it ever played.